src/ApacheOrcDotNet.OptimizedReader/Buffers/BaseColumnBuffer.cs

```diff
@@ -0,0 +1,285 @@
+        {
+            var stream = columnStreams.SingleOrDefault(stream =>
+                stream.StreamKind == streamKind
+            );
+
+            if (isRequired && stream == null)
+                throw new InvalidDataException($"The '{streamKind}' stream must be available");
+
+            return stream;
+        }
+
+        private protected void ReadByteStream(StreamDetail stream, ReadOnlySpan<byte> buffer, int length, Span<byte> outputValues, out int numValuesRead)
+        {
+            numValuesRead = 0;
+
+            if (stream == null)
+                return;
+
+            var numSkipped = 0;
+            var bufferReader = new BufferReader(ResizeBuffer(stream, buffer, length));
+
+            while (!bufferReader.Complete)
+            {
+                var numByteValuesRead = OptimizedByteRLE.ReadValues(ref bufferReader, _byteStreamBuffer);
+
+                for (int idx = 0; idx < numByteValuesRead; idx++)
+                {
+                    if (numSkipped++ < stream.Positions.ValuesToSkip)
+                        continue;
+
+                    outputValues[numValuesRead++] = _byteStreamBuffer[idx];
+
+                    if (numValuesRead >= outputValues.Length)
+                        return;
+                }
+            }
+        }
+
+        private protected void ReadBooleanStream(StreamDetail stream, ReadOnlySpan<byte> buffer, int length, Span<bool> outputValues, out int numValuesRead)
+        {
+            numValuesRead = 0;
+
+            if (stream == null)
+                return;
+
+            var numSkipped = 0;
+            var bufferReader = new BufferReader(ResizeBuffer(stream, buffer, length));
+            var numOfTotalBitsToSkip = stream.Positions.ValuesToSkip * 8 + stream.Positions.RemainingBits;
+            var numOfBytesToSkip = numOfTotalBitsToSkip / 8;
+            while (!bufferReader.Complete)
+            {
+                var numByteValuesRead = OptimizedByteRLE.ReadValues(ref bufferReader, _boolStreamBuffer);
+
+                for (int idx = 0; idx < numByteValuesRead; idx++)
+                {
+                    if (numSkipped++ < numOfBytesToSkip)
+                        continue;
+
+                    var decodedByte = _boolStreamBuffer[idx];
+                    var isFinalByte = bufferReader.Complete && idx >= numByteValuesRead - 1;
+
+                    // Skip remaining bits.
+                    if (numOfBytesToSkip % 8 != 0)
+                        decodedByte = (byte)(decodedByte << numOfTotalBitsToSkip % 8);
+
+                    if (isFinalByte && decodedByte == 0)
+                    {
+                        // Edge case where there is only one value for the row entry and that value is null
+                        outputValues[numValuesRead++] = false;
+                        return;
+                    }
+
+                    for (int bitIdx = 7; bitIdx >= 0; bitIdx--)
```

Comment:
The thought around simplifying this was that the compiler would be capable of optimizing this and possibly unrolling the loop, itself.

In terms of performance, it does not appear to have had a negative impact after simplifying. But it is simple to revert it back, if necessary.

Reply:
Could you create a little microbenchmark?  It would be good to have in our heads for future code if we need to be hand-optimizing stuff like this.

Reply:
Results for a single iteration:
|       Method |     Mean |     Error |    StdDev | Rank | Allocated |
|------------- |---------:|----------:|----------:|-----:|----------:|
|   NormalLoop | 9.014 ns | 0.0452 ns | 0.0401 ns |    2 |         - |
| UnrolledLoop | 2.451 ns | 0.0643 ns | 0.1176 ns |    1 |         - |

Results for 1000 iterations:
|       Method |      Mean |     Error |    StdDev | Rank | Allocated |
|------------- |----------:|----------:|----------:|-----:|----------:|
|   NormalLoop | 11.390 us | 0.0548 us | 0.0512 us |    2 |         - |
| UnrolledLoop |  2.666 us | 0.0216 us | 0.0202 us |    1 |         - |

I think that may be worth reverting it back. Any thoughts?

Reply:
I dug into this just a bit, as we expected the JIT to do a better job than it did.

Loop unrolling was well-supported in the old .Net Framework JIT, but has not been seriously implemented in the new JIT tooling that's written in C# (RyuJIT).  Here's the current assembly of your benchmarks:
![image](https://user-images.githubusercontent.com/1297262/179597919-c33f8705-cf02-424c-84a0-88396073f2ef.png)

We can see the loop is not being unrolled, as we predicted from the benchmark.  Notably, a lot of work is still happening in the unrolled version, that I think you could improve.

When you write a for loop like this:
```csharp
for(int i=0;i<array.Count;i++)
  doSomething(array[i]);
```
the JIT elides the range checks on `array[i]` since it knows the array access will be in bounds given the range restriction of the for loop.

You loop uses a variable external to the loop, and so we see all those checks against `0x1f40` (8000) to ensure the index isn't running off the end of the array.

This range check elision should work even if you're adding a constant value to the index, as static analysis should be able to determine that you won't run off the end.  But I'd drop that in SharpLab just to check.

Here's where .Net stands on loop unrolling:
Current State: https://github.com/dotnet/runtime/issues/4248
Future Work: https://github.com/dotnet/runtime/issues/65342

Reply:
Hmm, I see you're writing to a different sized array than the incoming array.  Let me think about that...

Reply:
So if you check the array (or Span) length before writing to it, the JIT will use that single check rather than verifying each time.
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgFRlwxoG8aG/++9JigYBVAHZQIAG2kwAJgBkIEAA4AKYAE8MMBvJiQD8gEI6YASl4D+PajYcMAbtigNxAV3wA1bNI/sAEow2PIMALwMAAwA3NaO/C5unvhmur7+QSFhkeRR+XH2CfwAyqrY4gA8wCrSAHwMEB4Yqs0ZAbgRDBzYYADWftIQYAw1MgDaKWkw7VmhDABUDAAcALqF8cUAZtAM6gCW4hgM+/IIXbEnZwyV7l7Ts7jBoTFXCADU71ZFxQxcv3xNgDGs1WhhHl0mi02n4OgA6ErSfZgGDqFKPZ7yCyFYECfZbdRQsGPOGKGDiADmGAAFjcVt9cY5gLBsH0Nj9cUSYZlcOMoqsuuoDEYFNMGAAyBjkUjLCwMACEkViQIBXPBsPY43IAsiQsMEGMYslqDliuiOMZIOh6p541IOr2woNovMEoYdFIpqVFsZase4zoDr1ItMrsl5CQXvNKt+fo1vJQQadhrD9IV3pjxTjtoArEn9SndG6UFHlRzgdmOuMkPmQ0aGJ709Hy6rQdyqwB2WvO0NF8Oln24zMJdHxzHvcLLQcJAC+mbn1BnQA===